MSAL/src/native_auth/public/state_machine/delegate_dispatcher/SignInDelegateDispatchers.swift

```diff
@@ -0,0 +1,132 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+
+final class SignInPasswordStartDelegateDispatcher: DelegateDispatcher<SignInPasswordStartDelegate> {
+
+    func dispatchSignInCodeRequired(
+        newState: SignInCodeRequiredState,
+        sentTo: String,
+        channelTargetType: MSALNativeAuthChannelType,
+        codeLength: Int
+    ) async {
+        if let onSignInCodeRequired = delegate.onSignInCodeRequired {
+            telemetryUpdate?(.success(()))
+            await onSignInCodeRequired(newState, sentTo, channelTargetType, codeLength)
+        } else {
+            let error = SignInPasswordStartError(type: .generalError, message: requiredErrorMessage(for: "onSignInCodeRequired"))
+            telemetryUpdate?(.failure(error))
+            await delegate.onSignInPasswordError(error: error)
+        }
+    }
+
+    func dispatchSignInCompleted(result: MSALNativeAuthUserAccountResult) async {
+        if let onSignInCompleted = delegate.onSignInCompleted {
+            telemetryUpdate?(.success(()))
+            await onSignInCompleted(result)
+        } else {
+            let error = SignInPasswordStartError(type: .generalError, message: requiredErrorMessage(for: "onSignInCompleted"))
+            telemetryUpdate?(.failure(error))
+            await delegate.onSignInPasswordError(error: error)
+        }
+    }
+}
+
+final class SignInStartDelegateDispatcher: DelegateDispatcher<SignInStartDelegate> {
+
+    func dispatchSignInCodeRequired(
+        newState: SignInCodeRequiredState,
+        sentTo: String,
+        channelTargetType: MSALNativeAuthChannelType,
+        codeLength: Int
+    ) async {
+        if let onSignInCodeRequired = delegate.onSignInCodeRequired {
```

Comment:
it seems to see a pattern here:
`if delegate.expectedMethod is implemented {
    telemetryUpdate?(.success(()))
    call expectedMethod
 } else {
      create error
      telemetryUpdate?(.failure(error))
      call delegate.error
} `
In the controller classes we used to use function that accepted a generic success and error delegate methods. Do you think is it possible to applies something similar here? We could save a lot of duplicate code.

Reply:
I did it but I don't like the result, it's too complex for the small logic we have in this case. In the controllers I agree, that it's a great choice because it saves us a lot of duplication, but here I'm not sure. This is how it looks like after having created two generic functions in the parent class `DelegateDispatcher`:

```swift
    func dispatchSignInCodeRequired(
        newState: SignInCodeRequiredState,
        sentTo: String,
        channelTargetType: MSALNativeAuthChannelType,
        codeLength: Int
    ) async {
        if let onSignInCodeRequired = delegate.onSignInCodeRequired {
            await completeDelegate { await onSignInCodeRequired(newState, sentTo, channelTargetType, codeLength) }
        } else {
            let error = SignInPasswordStartError(type: .generalError, message: requiredErrorMessage(for: "onSignInCodeRequired"))
            await completeDelegateError(error) { [weak self] in
                await self?.delegate.onSignInPasswordError(error: error)
            }
        }
    }
```

If you have a different thing in mind, let me know.

Reply:
I thought we could use something similar as we used in the SignInController, where we were passing a code block as method parameter (as Silviu is showing in his comment). But if it would result in an addition of too much complexity I'm happy to leave it as it is